packages/auth-remix/src/server.ts

```diff
@@ -0,0 +1,795 @@
+
+export interface CreateAuthRouteHandlers {
+  onOAuthCallback(
+    params: ParamsOrError<{
+      tokenData: TokenData;
+      provider: BuiltinOAuthProviderNames;
+      isSignUp: boolean;
+    }>
+  ): Promise<Response>;
+  onBuiltinUICallback(
+    params: ParamsOrError<
+      (
+        | {
+            tokenData: TokenData;
+            provider: BuiltinProviderNames;
+          }
+        | { tokenData: null; provider: null }
+      ) & { isSignUp: boolean }
+    >
+  ): Promise<Response>;
+  onEmailVerify(
+    params: ParamsOrError<
+      { tokenData: TokenData },
+      { verificationToken?: string }
+    >
+  ): Promise<Response>;
+  onSignout(): Promise<Response>;
+}
+
+export class RemixServerAuth extends RemixClientAuth {
+  private readonly client: Client;
+  private readonly core: Promise<Auth>;
+
+  constructor(client: Client, options: RemixAuthOptions) {
+    super(options);
+
+    this.client = client;
+    this.core = Auth.create(client);
+  }
+
+  isPasswordResetTokenValid(resetToken: string) {
+    return Auth.checkPasswordResetTokenValid(resetToken);
+  }
+
+  getSession(request: Request) {
+    return new RemixAuthSession(
+      this.client,
+      parseCookies(request)[this.options.authCookieName]
+    );
+  }
+
+  async getProvidersInfo() {
+    return (await this.core).getProvidersInfo();
+  }
+
+  createAuthRouteHandlers({
+    onOAuthCallback,
+    onBuiltinUICallback,
+    onEmailVerify,
+    onSignout,
+  }: Partial<CreateAuthRouteHandlers>) {
+    return {
+      loader: async ({
+        request,
+        params,
+      }: {
+        request: Request;
+        params: { [key: string]: string | undefined };
+      }) => {
+        const path = params["*"];
+        if (!path) {
+          throw new Error(`route handlers file should end with '.$.ts'`);
+        }
+        const searchParams = new URL(request.url).searchParams;
+        switch (path) {
+          case "oauth": {
+            if (!onOAuthCallback) {
+              throw new Error(
+                `'onOAuthCallback' auth route handler not configured`
+              );
+            }
+            const provider = searchParams.get(
+              "provider_name"
+            ) as BuiltinOAuthProviderNames | null;
+
+            if (!provider || !builtinOAuthProviderNames.includes(provider)) {
+              throw new Error(`invalid provider_name: ${provider}`);
+            }
+            const redirectUrl = `${this._authRoute}/oauth/callback`;
+            const pkceSession = await this.core.then((core) =>
+              core.createPKCESession()
+            );
+
+            return redirect(
+              pkceSession.getOAuthUrl(
+                provider,
+                redirectUrl,
+                `${redirectUrl}?isSignUp=true`
+              ),
+              {
+                headers: new Headers({
+                  "Set-Cookie": `${this.options.pkceVerifierCookieName}=${pkceSession.verifier}; HttpOnly;`,
```

Comment:
Maybe we should use a cookie lib like this one: https://www.npmjs.com/package/cookie, instead of manually building the strings here?